File: gfg/medium/longest-sub-array-with-sum-k0809/solution.cpp
class Solution {
  public:
    int longestSubarray(vector<int>& arr, int k) {
        unordered_map<long long, int> mp;
        long long prefixSum = 0;
        int maxLen = 0;

        for (int i = 0; i < arr.size(); i++) {
            prefixSum += arr[i];

            // Subarray from index 0 to i
            if (prefixSum == k) {
                maxLen = i + 1;
            }

            // Check if (prefixSum - k) exists
            if (mp.find(prefixSum - k) != mp.end()) {
                maxLen = max(maxLen, i - mp[prefixSum - k]);
            }

            // Store first occurrence of prefixSum
            if (mp.find(prefixSum) == mp.end()) {
                mp[prefixSum] = i;
            }
        }

        return maxLen;
    }
};

File: gfg/medium/longest-sub-array-with-sum-k0809/solution.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @returns {number}
 */
class Solution {
    longestSubarray(arr, k) {
        let map = new Map();
        let sum = 0;
        let maxLen = 0;

        for (let i = 0; i < arr.length; i++) {
            sum += arr[i];

            // Subarray from index 0 to i
            if (sum === k) {
                maxLen = i + 1;
            }

            // Check if prefix sum (sum - k) exists
            if (map.has(sum - k)) {
                maxLen = Math.max(maxLen, i - map.get(sum - k));
            }

            // Store first occurrence only
            if (!map.has(sum)) {
                map.set(sum, i);
            }
        }

        return maxLen;
    }
}